src/device/asgard_kronmult_cpu_general.hpp

```diff
@@ -0,0 +1,416 @@
+        serialized(n *
+                   ((dimensions == 6)
+                        ? n5
+                        : ((dimensions == 5)
+                               ? n4
+                               : ((dimensions == 4)
+                                      ? n3
+                                      : ((dimensions == 3)
+                                             ? n2
+                                             : ((dimensions == 2) ? n : 1))))))
+  {
+    static_assert(1 <= dimensions and dimensions <= 6);
+  }
+  T &operator()(int i, int j, int k, int l, int m, int p)
+  {
+    static_assert(dimensions == 6);
+    return serialized[i * n5 + j * n4 + k * n3 + l * n2 + m * n + p];
+  }
+  T &operator()(int j, int k, int l, int m, int p)
+  {
+    static_assert(dimensions == 5);
+    return serialized[j * n4 + k * n3 + l * n2 + m * n + p];
+  }
+  T &operator()(int k, int l, int m, int p)
+  {
+    static_assert(dimensions == 4);
+    return serialized[k * n3 + l * n2 + m * n + p];
+  }
+  T &operator()(int l, int m, int p)
+  {
+    static_assert(dimensions == 3);
+    return serialized[l * n2 + m * n + p];
+  }
+  T &operator()(int m, int p)
+  {
+    static_assert(dimensions == 2);
+    return serialized[m * n + p];
+  }
+  T &operator()(int p)
+  {
+    static_assert(dimensions == 1);
+    return serialized[p];
+  }
+  void zero()
+  {
+    for (auto &s : serialized)
+      s = 0;
+  }
+  size_t size() const { return serialized.size(); }
+
+private:
+  int const n5, n4, n3, n2, n;
+  std::vector<T> serialized;
+};
+
+template<typename T, int dimensions, scalar_case alpha_case,
+         scalar_case beta_case>
+void cpu_dense(int const n, int const num_rows, int const num_cols,
+               int const num_terms, int const iA[], T const vA[], T const alpha,
+               T const x[], T const beta, T y[])
+{
+#pragma omp parallel
+  {
+    tensor<T, dimensions> Y(n), W(n);
+
+// always use one thread per kron-product
+#pragma omp for
+    for (int iy = 0; iy < num_rows; iy++)
+    {
+      // tensor i (ti) is the first index of this tensor in y
+      int const ti = iy * Y.size();
+      if constexpr (beta_case == scalar_case::zero)
+        for (size_t j = 0; j < Y.size(); j++)
+          y[ti + j] = 0;
+      else if constexpr (beta_case == scalar_case::neg_one)
+        for (size_t j = 0; j < Y.size(); j++)
+          y[ti + j] = -y[ti + j];
+      else if constexpr (beta_case == scalar_case::other)
+        for (size_t j = 0; j < Y.size(); j++)
+          y[ti + j] *= beta;
+
+      // ma is the starting index of the operators for this y
+      int ma = iy * num_cols * num_terms * dimensions;
+
+      for (int jx = 0; jx < num_cols; jx++)
+      {
+        // tensor i (ti) is the first index of this tensor in x
+        int const tj = jx * Y.size();
+        for (int t = 0; t < num_terms; t++)
+        {
+          if constexpr (dimensions == 1)
+          {
+            Y.zero();
+            T const *A = &(vA[iA[ma++]]);
```

Comment:
```suggestion
            T const * const A = &(vA[iA[ma++]]);
```
and similar throughout this file